libraries/AP_AIS/AP_AIS.cpp

```diff
@@ -0,0 +1,798 @@
+                return;
+        }
+    }
+}
+
+// removed the given index from the AIVDM buffer shift following elements
+void AP_AIS::buffer_shift(uint8_t i)
+{
+    for (uint8_t n = i;  n < (AIVDM_BUFFER_SIZE - 1); n++) {
+        _AIVDM_buffer[n].ID = _AIVDM_buffer[n+1].ID;
+        _AIVDM_buffer[n].num = _AIVDM_buffer[n+1].num;
+        _AIVDM_buffer[n].total = _AIVDM_buffer[n+1].total;
+        strncpy(_AIVDM_buffer[n].payload,_AIVDM_buffer[n+1].payload,65);
+    }
+    _AIVDM_buffer[AIVDM_BUFFER_SIZE - 1].ID = 0;
+    _AIVDM_buffer[AIVDM_BUFFER_SIZE - 1].num = 0;
+    _AIVDM_buffer[AIVDM_BUFFER_SIZE - 1].total = 0;
+    _AIVDM_buffer[AIVDM_BUFFER_SIZE - 1].payload[0] = 0;
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Functions related to the vessel list
+
+// find vessel index in existing list, if not then return new index if possible
+bool AP_AIS::get_vessel_index(uint32_t mmsi, uint16_t &index, uint32_t lat, uint32_t lon)
+{
+    const uint16_t list_size = _list.max_items();
+
+    uint16_t empty = 0;
+    bool found_empty = false;
+    for (uint16_t i = 0; i < list_size; i++) {
+        if (_list[i].info.MMSI == mmsi) {
+            index = i;
+            return true;
+        }
+        if (_list[i].last_update_ms == 0 && !found_empty) {
+                found_empty = true;
+                empty = i;
+        }
+    }
+
+    // got through the list without a match
+    if (found_empty) {
+        index = empty;
+        _list[index].info.MMSI = mmsi;
+        return true;
+    }
+
+    // no space in the list
+    if (list_size < _max_list) {
+        // if we can try and expand
+        if (_list.expand(1)) {
+            index = list_size;
+            _list[index].info.MMSI = mmsi;
+            return true;
+        }
+    }
+
+    // could not expand list, either because of memory or max list param
+    // if we have a valid incoming location we can bump a further item from the list
+    if (lat == 0 && lon ==0) {
+        return false;
+    }
+
+    struct Location current_loc;
+    if (!AP::ahrs().get_position(current_loc)) {
+        return false;
+    }
+
+    struct Location loc;
+    float dist;
+    float max_dist = 0;
+    for (uint16_t i = 0; i < list_size; i++) {
+        loc.lat = _list[i].info.lat;
+        loc.lng = _list[i].info.lon;
+        dist = loc.get_distance(current_loc);
+        if (dist > max_dist) {
+            max_dist = dist;
+            index = i;
+        }
+    }
+
+    // find the curent distance
+    loc.lat = lat;
+    loc.lng = lon;
+    dist = loc.get_distance(current_loc);
+
+    if (dist < max_dist) {
+        clear_list_item(index);
+        _list[index].info.MMSI = mmsi;
+        return true;
+    }
+
+    return false;
+}
+
+void AP_AIS::clear_list_item(uint16_t index)
+{
+    if (index < _list.max_items()) {
+        _list[index].last_update_ms = 0;
```

Comment:
i wonder if there's a more efficient way to set these all to zero if we know that every item in the structure should be set to zero.  A memset is one idea..